File: backend/src/cuanto_cuesta/application/limits.py
"""The caps on fee values, for defaults and user overrides alike.

They reject typos and nonsense (a 90 % fee, a 5000 USD fixed fee), not unusual but real fees:
every default in ``fees.yaml`` sits far below them.
"""

from __future__ import annotations

from decimal import Decimal

from cuanto_cuesta.domain import Currency, Fee, FixedFee, PercentFee

MAX_PERCENT = Decimal(20)


def max_fixed(currency: Currency) -> Decimal:
    """The cap on a fixed fee or a percent fee minimum in ``currency``."""
    match currency:
        case Currency.USD | Currency.USDT | Currency.USDC:
            return Decimal(100)
        case Currency.ARS:
            return Decimal(150_000)


def value_problem(value: Decimal, cap: Decimal) -> str | None:
    """Why ``value`` is not an acceptable fee value, or None if it is."""
    if not value.is_finite():
        return f"must be a finite number, got {value}"
    if value < 0:
        return f"must not be negative, got {value}"
    if value > cap:
        return f"must be at most {cap}, got {value}"
    return None


def fee_problems(fee: Fee) -> list[str]:
    """Why an already built fee is outside the caps, if it is."""
    match fee:
        case FixedFee(amount=amount):
            checks = [("value", amount.amount, max_fixed(amount.currency))]
        case PercentFee(rate=rate, minimum=minimum):
            checks = [("value", rate.value, MAX_PERCENT)]
            if minimum is not None:
                checks.append(("minimum", minimum.amount, max_fixed(minimum.currency)))
    return [
        f"{field} {problem}"
        for field, value, cap in checks
        if (problem := value_problem(value, cap)) is not None
    ]

File: frontend/src/calculator/limits.ts
/**
 * The caps on fee values, for defaults and user edits alike. Mirrors
 * `backend/src/cuanto_cuesta/application/limits.py`.
 *
 * They reject typos and nonsense (a 90 % fee, a 5000 USD fixed fee), not unusual but real fees.
 * Problems are codes, not sentences: the screen writes the message.
 */

import type { Fee } from "./fees.ts";
import { Decimal, type Big, type Currency } from "./money.ts";

export const MAX_PERCENT = new Decimal(20);

export function maxFixed(currency: Currency): Big {
  switch (currency) {
    case "USD":
    case "USDT":
    case "USDC":
      return new Decimal(100);
    case "ARS":
      return new Decimal(150_000);
  }
}

export type ValueProblem =
  { readonly code: "negative" } | { readonly code: "above_cap"; readonly cap: Big };

export function valueProblem(value: Big, cap: Big): ValueProblem | null {
  if (value.lt(0)) {
    return { code: "negative" };
  }
  if (value.gt(cap)) {
    return { code: "above_cap", cap };
  }
  return null;
}

/** The cap for a fee's main value: a percentage or an amount in the fee's currency. */
export function valueCap(fee: Fee): Big {
  switch (fee.kind) {
    case "fixed":
      return maxFixed(fee.amount.currency);
    case "percent":
      return MAX_PERCENT;
  }
}

export interface FeeProblem {
  readonly field: "value" | "minimum";
  readonly problem: ValueProblem;
}

/** Every value of an already built fee that is outside the caps, the minimum included. */
export function feeProblems(fee: Fee): FeeProblem[] {
  const problems: FeeProblem[] = [];
  const value = fee.kind === "fixed" ? fee.amount.amount : fee.rate;
  const valueCheck = valueProblem(value, valueCap(fee));
  if (valueCheck !== null) {
    problems.push({ field: "value", problem: valueCheck });
  }
  if (fee.kind === "percent" && fee.minimum !== null) {
    const minimumCheck = valueProblem(fee.minimum.amount, maxFixed(fee.minimum.currency));
    if (minimumCheck !== null) {
      problems.push({ field: "minimum", problem: minimumCheck });
    }
  }
  return problems;
}
